hermes/src/store.rs

```diff
@@ -1,73 +1,212 @@
+            ),
+            pending_accumulations: Cache::builder()
+                .max_capacity(10_000)
+                .time_to_live(Duration::from_secs(60 * 5))
+                .build(), // FIXME: Make this configurable
         }
     }
 
-    /// Stores the update data in the store and returns the price identifiers for which
-    /// price feeds were updated.
-    pub fn store_update(&self, update: Update) -> Result<Vec<PriceIdentifier>> {
-        match update {
+    /// Stores the update data in the store
+    pub async fn store_update(&self, update: Update) -> Result<()> {
+        let ring_index = match update {
             Update::Vaa(vaa_bytes) => {
-                proof::batch_vaa::store_vaa_update(self.state.clone(), vaa_bytes)
+                let vaa = VAA::from_bytes(vaa_bytes.clone())?;
+                let payload = WormholePayload::try_from_bytes(&vaa.payload, &vaa_bytes)?;
+
+                // FIXME: Validate the VAA
+                // FIXME: Skip similar VAAs
+
+                match payload {
+                    WormholePayload::Merkle(proof) => {
+                        log::info!("Storing merkle proof for state index {:?}", proof);
+                        store_wormhole_merkle_verified_message(self, proof.clone()).await?;
+                        proof.state_index
+                    }
+                }
             }
-        }
+            Update::AccumulatorMessages(accumulator_messages) => {
+                // FIXME: Move this constant to a better place
+                const RING_SIZE: u32 = 10_000;
+                let ring_index = accumulator_messages.slot.rem(RING_SIZE as u64) as u32;
+
+                log::info!(
+                    "Storing accumulator messages for ring index {:?}: {:?}",
+                    ring_index,
+                    accumulator_messages
+                );
+
+                let pending_acc = self
+                    .pending_accumulations
+                    .entry(ring_index)
+                    .or_default()
+                    .await
+                    .into_value();
+                self.pending_accumulations
+                    .insert(
+                        ring_index,
+                        pending_acc.accumulator_messages(accumulator_messages),
+                    )
+                    .await;
+
+                ring_index
+            }
+        };
+
+        let pending_state = self.pending_accumulations.get(&ring_index);
+        let pending_state = match pending_state {
+            Some(pending_state) => pending_state,
+            // Due to some race conditions this might happen when it's processed before
+            None => return Ok(()),
+        };
+
+        let state = match pending_state.build() {
+            Ok(state) => state,
+            Err(_) => return Ok(()),
+        };
+
+        log::info!("State: {:?}", state);
+
+        let wormhole_merkle_message_states_proofs = construct_message_states_proofs(state.clone())?;
```

Comment:
I thought about it. There are some tradeoffs here. Most importantly, I'll like the storage to have this proof per message completely because it will probably gets stored off-chain and will be easier for reconstruction. Making it lazy will make the storage api more complex.